src/swarm-scenario/Swarm/Game/Scenario/Topography/Structure/Overlay.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- |
+-- SPDX-License-Identifier: BSD-3-Clause
+--
+-- Generic overlay operations on grids
+module Swarm.Game.Scenario.Topography.Structure.Overlay (
+  PositionedGrid (..),
+) where
+
+import Control.Applicative
+import Data.Function (on)
+import Data.Int (Int32)
+import Data.Tuple (swap)
+import Linear
+import Swarm.Game.Location
+import Swarm.Game.Scenario.Topography.Area
+import Swarm.Util (applyWhen)
+
+data PositionedGrid a = PositionedGrid
+  { gridPosition :: Location
+  , gridContent :: Grid a
+  }
+  deriving (Eq)
+
+instance Show (PositionedGrid a) where
+  show (PositionedGrid p g) =
+    unwords
+      [ "Grid with dimension"
+      , renderRectDimensions $ getGridDimensions g
+      , "located at"
+      , show p
+      ]
+
+data OverlayPair a = OverlayPair
+  { _base :: a
+  , _overlay :: a
+  }
+
+getBottomRightCorner :: PositionedGrid a -> Location
```

Comment:
Seems like this code (`getBottomRightCorner`, `getNorthwesternExtent`, `getSoutheasternExtent`, etc.) could be made simpler by introducing some kind of `Rectangle` abstraction which consists of two `Location`s, an upper left and a bottom right (or perhaps a `Location` + `AreaDimension`).  Then define a `Semigroup` instance on `Rectangles` which takes the max + min appropriately, i.e. `r1 <> r2` is the smallest rectangle that contains both `r1` and `r2`.  Then instead of `getBottomRightCorner` we can just have a `getRectangle` function which gets the `Rectangle` corresponding to a `PositionedGrid`, etc.

Reply:
@byorgey added a `SubsumingRect` for this.